// nexmo-cli/lib/client.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _nexmo = require('nexmo');

var _nexmo2 = _interopRequireDefault(_nexmo);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Client = function () {
  function Client(config, emitter) {
    _classCallCheck(this, Client);

    this.emitter = emitter;
    this.config = config;
  }

  _createClass(Client, [{
    key: 'instance',
    value: function instance() {
      var nexmo = initialize(this.config, this.emitter);
      return nexmo;
    }
  }, {
    key: 'definition',
    value: function definition() {
      return _nexmo2.default;
    }
  }, {
    key: 'instanceWith',
    value: function instanceWith(key, secret) {
      return new _nexmo2.default({ apiKey: key, apiSecret: secret });
    }
  }]);

  return Client;
}();

exports.default = Client;

// private methods

var initialize = function initialize(config, emitter) {
  var nexmo = null;
  var packageDetails = require(__dirname + '/../package.json');

  try {
    var credentials = config.read().credentials;
    nexmo = new _nexmo2.default({
      apiKey: credentials.api_key,
      apiSecret: credentials.api_secret
    }, {
      debug: emitter.debugging,
      appendToUserAgent: 'nexmo-cli/' + packageDetails.version.replace('v', '')
    });
  } catch (e) {
    if (e instanceof TypeError) {
      emitter.error('Could not initialize Nexmo library. Please run \'nexmo setup\' to setup the CLI correctly. (' + e.message + ')');
    } else {
      emitter.error('Could not read credentials. Please run \'nexmo setup\' to setup the CLI. (' + e.message + ')');
    }
  }
  return nexmo;
};
